server/src/main/java/org/elasticsearch/gateway/GatewayMetaState.java

```diff
@@ -109,7 +109,11 @@ public void start(Settings settings, TransportService transportService, ClusterS
                             .build());
                     lucenePersistedState = new LucenePersistedState(
                         persistenceWriter, currentTerm, clusterState);
-                    metaStateService.deleteAll(); // delete legacy files
+                    if (DiscoveryNode.isDataNode(settings)) {
+                        metaStateService.unreferenceAll(); // unreference legacy files (only keep them for dangling indices functionality)
```

Comment:
:+1: noting that if dangling indices functionality is disabled then we delete the metadata as the indices are allocated, so if any indices are genuinely dangling at this point then we keep their metadata around.

Reply:
yup.